### PR TITLE
very first interim_l3 dataset

### PR DESCRIPTION
**Pull request by hgloeckner**
I added a first draft for a gridded class that can be used to start l3 creation. 
Right now it does some things where I am not sure if that's useful: 
- `launch_time` is made a dimension 
- datasets are concatenated along `launch_time` and `time`causing many missing values (those will go away when we switch dimension to some `height` in subsequent steps, but maybe the switch should happen before?)
- all the former attributes (aircraft data, like `true_heading_(deg)`, but also things like `author`) are now variables. We should probably add something that looks at them and if it's the same thing for every sonde makes it an attribute again. I just did not want to lose information. 

It's not meant to be perfect at this point, but maybe we can use it to discuss whether it makes sense like this at all?

### CODE REVIEW

**Comment by hgloeckner** [comment]
... and Issue #120 should be solved here as well - for now I just copied the Sonde header and way to do things

**Comment by Geet-George** [comment]
Hi @hgloeckner, thanks for starting this. I think it makes a lot of sense, but maybe with some changes? I'm "typing out loud" my thoughts here. 

There are two main steps (and a whole bunch of sub-steps, see #90) to go from L2 to L3:

(a) Prepare L2 data for concatenation (interpolate, cell methods, etc.) - this still happens to `Sonde` instances. The final step could be your function `prepare_l2_for_gridded` of keeping attributes, add `launch_time` dim, etc. (although `sonde_id` is the dim we want for `Gridded` not `launch_time`.)

(b) Concatenate & incorporate L3 features (attrs, vars, etc.) - this happens to `Gridded` instances

After (a) is done, this will be stored in an attribute (currently `_interim_l3_ds`, we could probably change the name!) I see the `Gridded` class as something that would concatenate multiple instances from `Sonde` (in particular the attribute `_interim_l3_ds`) and then apply further changes, maybe such as adding `wspd`,`wdir`,`T`,`RH`, other attributes, etc. 

To summarize, there will be a bunch of functions between "add_l2_ds" and "prepare_l2_for_gridded" in the pipeline, and we should make `sonde_id` the dimension for gridded sondes, instead of `launch_time`.

I hope these make sense. Otherwise the PR is exactly in the direction where we want to go! :)

File: src/halodrops/processor.py [hunk 1]
```diff
@@ -967,9 +986,32 @@ def add_q_and_theta_to_l2_ds(self):
...
 
         return self
+
+
+@dataclass(order=True, frozen=True)
+class Gridded:
+    sondes: dict
+    # platforms: dict = field(default_factory=dict)
+    _: KW_ONLY
+    interpolation_grid: np.ndarray = field(
```
Comment (Geet-George) [inline src/halodrops/processor.py:1002]: These class attributes of `interpolation_grid`, `interpolation_bins` and `max_gap_fill` will make it fixed values, and unable to be changed by user config. I think it is better to have these values as something that can be imported from a `helper` function. The `concat_sondes` can then use these as arguments.

File: src/halodrops/processor.py [hunk 2]
```diff
@@ -967,9 +986,32 @@ def add_q_and_theta_to_l2_ds(self):
...
+    interpolation_bins: np.ndarray = field(
+        default_factory=lambda: np.arange(-5, 10015, 10).astype("int")
+    )
+    max_gap_fill: int = 50
+
+    def concat_sondes(self):
+        """
+        function to concatenate all sondes using the combination of all measurement times and launch times
+        """
+        list_of_l2_ds = [sonde.l2_ds for sonde in self.sondes.values()]
```
Comment (Geet-George) [inline src/halodrops/processor.py:1014]:
```suggestion
        list_of_l2_ds = [sonde._interim_l3_ds for sonde in self.sondes.values()]
```

The attribute `l2_ds` should stay unchanged because that will be the access to L2 data. Instead anything modified on that can be in the attr `_interim_l3_ds`, which will be mutable and the eventual state of that attribute should be what `Gridded` wants to concatenate.

File: src/halodrops/processor.py [hunk 3]
```diff
@@ -967,9 +986,32 @@ def add_q_and_theta_to_l2_ds(self):
         else:
             ds = self.l2_ds
 
-        ds = calc_q_from_rh(ds)
-        ds = calc_theta_from_T(ds)
+        ds = hh.calc_q_from_rh(ds)
```
Comment (Geet-George) [inline src/halodrops/processor.py:989]: This should be a different commit. (that's my bad to not include `hh`, sorry :P)

File: src/halodrops/processor.py [hunk 4]
```diff
@@ -949,6 +949,25 @@ def add_l2_ds(self, l2_dir: str = None):
...
+        """
+        l2_ds = self.l2_ds.assign_coords(
+            {"launch_time": ("launch_time", [self.launch_time])}
+        ).sortby("time")
+
+        for attr, value in self.l2_ds.attrs.items():
+            l2_ds[attr] = value
+
+        l2_ds.attrs.clear()
+        object.__setattr__(self, "l2_ds", l2_ds)
```
Comment (Geet-George) [inline src/halodrops/processor.py:968]:
```suggestion
        object.__setattr__(self, "_interim_l3_ds", l2_ds)
```

File: src/halodrops/processor.py [hunk 5]
```diff
@@ -949,6 +949,25 @@ def add_l2_ds(self, l2_dir: str = None):
...
 
+    def prepare_l2_for_gridded(self):
+        """
+        Prepares l2 datasets to be concatenated to gridded.
+        I.e adds launch_time as dimension
+        adds all attributes as variables to avoid conflicts when concatenating because attributes are different
+        (and not lose information)
+
+        """
+        l2_ds = self.l2_ds.assign_coords(
```
Comment (Geet-George) [inline src/halodrops/processor.py:960]:
```suggestion
        _interim_l3_ds = self.l2_ds.assign_coords(
```

File: src/halodrops/processor.py [hunk 6]
```diff
@@ -949,6 +949,25 @@ def add_l2_ds(self, l2_dir: str = None):
...
+            {"launch_time": ("launch_time", [self.launch_time])}
+        ).sortby("time")
+
+        for attr, value in self.l2_ds.attrs.items():
+            l2_ds[attr] = value
```
Comment (Geet-George) [inline src/halodrops/processor.py:965]:
```suggestion
            _interim_l3_ds[attr] = value
```

File: src/halodrops/processor.py [hunk 7]
```diff
@@ -949,6 +949,25 @@ def add_l2_ds(self, l2_dir: str = None):
...
+
+        l2_ds.attrs.clear()
```
Comment (Geet-George) [inline src/halodrops/processor.py:967]:
```suggestion
        _interim_l3_ds.attrs.clear()
```

File: src/halodrops/processor.py (hunk 3 above)
Comment (hgloeckner) [reply src/halodrops/processor.py:989]: yes true. I subcontiously changed it at some point ^^'

**Comment by hgloeckner** [comment]
Your completely right - Nina, Theresa and I will also talk this afternoon. 
Interpolating before is definitely more useful. I unfortunately only realized that too late and wanted to already give something up for discussion^^' I hope I can fix your points later today

File: src/halodrops/pipeline.py [hunk 8]
```diff
@@ -272,8 +272,11 @@ def iterate_Sonde_method_over_dict_of_Sondes_objects(
     return my_dict
 
 
-def sondes_to_gridded(sondes: dict) -> xr.Dataset:
-    pass
+def sondes_to_gridded(sondes: dict, config: configparser.ConfigParser):
+    print(sondes)
```
Comment (Geet-George) [inline src/halodrops/pipeline.py:276]:
```suggestion
```

File: src/halodrops/processor.py [hunk 9]
```diff
@@ -949,6 +949,65 @@ def add_l2_ds(self, l2_dir: str = None):
...
+    def create_interim_l3(self):
+        _interim_l3_ds = self.l2_ds.assign_coords(
+            {"serial_id": ("serial_id", [self.serial_id])}
+        ).sortby("time")
+        object.__setattr__(self, "_interim_l3_ds", _interim_l3_ds)
+        return self
+
+    def remove_non_mono_incr_alt(self):
+        """
+        This function removes the indices in the
```
Comment (Geet-George) [inline src/halodrops/processor.py:961]:
```suggestion
        This function removes the value at the indices which have a non-monotonically increasing value in the
```

**Review by Geet-George — COMMENTED** [review]
I wasn't able to test it out, but from looking at the code, it looks okay. I think something might break if we don't provide a file that is prepared from L2 before concatenating, but this is worth trying when we actually add in the L3 functions.

**Comment by hgloeckner** [comment]
I am not really sure if I accidentally deleted something essential from the interpolation method compared to joanne. One downside in the comparison is that my version needs the `bottleneck` package to run (I think otherwise interpolate_na over a time variable does not work out of the box). I also haven't thoroughly checked the functions. so it's not ready for merge yet, but maybe you can have a look if you like

**Comment by hgloeckner** [comment]
It's not working if there are sondes without a launch detect yet. Fixing that currently by removing those from the sonde_dictionary. That means that the sondes that don't detect a launch are removed from the dataset, but I guess for level3 that's ok (they still show in level2). Also, are the sondes where the qc fails currently still in there? then we should remove them as well

**Comment by hgloeckner** [comment]
I get a basic level3 file out of this. However, I am not sure whether we are (correctly) iterating over the different flight IDs. As far as I understood there should be a level3 file for each flight right? I guess we are not doing that currently

**Comment by hgloeckner** [comment]
I added the q and theta calculations since I needed them for the plot :)

**Comment by ninarobbins** [comment]
hi @hgloeckner, when trying to plot the sondes in a map, I realized that the lon/lat have dimension (gpsalt), when they should also have dimension (sonde_id). I'll try to fix this

**Comment by hgloeckner** [comment]
I think also, main is not working anymore right now (or only for one path structure) because the last to commits are actually fixes for the path_structure not so much for this branch.

**Comment by hgloeckner** [comment]
> I think also, main is not working anymore right now (or only for one path structure) because the last to commits are actually fixes for the path_structure not so much for this branch.

I put those changes into another pull request so that we can have a working main before we merge this :)

**Comment by ninarobbins** [comment]
Where does the sonde_id currently come from? I guess it should be something containing the platform, flight date, and sonde number (HALO_20240811_s1).

**Comment by hgloeckner** [comment]
It's the internal serial id. Geet and Tobi answered this to me in Issue #127

**Comment by Geet-George** [comment]
> Are you okay with merging? I think it would be nice to have a basic level 3 in main so that we can start further editing from there @Geet-George

I am not sure I could completely understand the conversation here. So, is the main branch working now? And what does this basic level 3 include?
